engine/framework/base_jobmaster.go

```diff
@@ -216,16 +214,38 @@ func (d *DefaultBaseJobMaster) GetWorkers() map[frameModel.WorkerID]WorkerHandle
 
 // Close implements BaseJobMaster.Close
 func (d *DefaultBaseJobMaster) Close(ctx context.Context) error {
-	err := d.impl.CloseImpl(ctx)
-	// We don't return here if CloseImpl return error to ensure
-	// that we can close inner resources of the framework
-	if err != nil {
-		log.L().Error("Failed to close JobMasterImpl", zap.Error(err))
-	}
+	d.closeOnce.Do(func() {
+		err := d.impl.CloseImpl(ctx)
+		if err != nil {
+			log.L().Error("Failed to close JobMasterImpl", zap.Error(err))
+		}
+	})
```

Comment:
If `closeOnce` and `CloseImpl` is strongly binding, may be we can wrap them together.